### ISSUE TITLE
C API for Symmetric encryption: multiple encryption calls with the same secret/password

### ISSUE THREAD

**Issue by jpo234** [rnpgp/rnp#991]
## Description

It seems that the rnp_op_encrypt_t can't be used for multiple consecutive calls with the same password. That's bad, because the key derivation functions are very expensive on purpose (many iterations).
So, what I'm looking for is a way to initialize a rnp_op_encrypt_t once and then do multiple rnp_op_encrypt_execute calls. Alternatively there should be a way to precompute or cache the key without the need to call the KDF again for every encryption call.

**Comment by ni4** [comment]
@jpo234 yeah, we don't have KDF caching yet. But why and where do you need it?

**Comment by jpo234** [comment]
@ni4 I'm working on a legacy application that sends numerous small messages over a slow radio modem (really slow, like forget-a-VPN-9600-bps). The aim is to encrypt the messages that are currently sent in plain text. Instead of badly reinventing an encrypted message format the design document wisely uses the openpgp format.
Just now I'm evaluating the available options and librnp really looks very promising, but has a few warts, like the one we are just talking about.

**Comment by ni4** [comment]
@jpo while it's not hard to implement password precomputing I doubt that it is good from the security perspective, since same password + salt will be used for small messages encryption.
How many messages do you need to encrypt per second? Maybe some Ed25519 or ECDH public key encryption/decryption will work fast enough.

**Comment by jpo234** [comment]
I just looked that up. The key gets indeed calculated from the secret and the salt. I assumed that the salt gets prepended to the plain text.
So, yes, I agree that using the same key for multiple messages is insecure. Too bad.

As for your suggestion regarding a public key algorithm: I tried this initially with RSA. The problem is that the messages became 5x as large (~500 bytes instead of ~100). I'll just suck it up and recalculate the key every time.

**Comment by ni4** [comment]
@jpo234 RSA has much larger overhead (around the key's size, i.e. 500+ bytes for 4096 bit key. ECDH/Curve25519 should have overhead less then 100 bytes as far as I remember.
